system/lib/standalone/standalone.c

```diff
@@ -65,21 +104,106 @@ weak int _munmap_js(
   return -ENOSYS;
 }
 
-// open(), etc. - we just support the standard streams, with no
-// corner case error checking; everything else is not permitted.
-// TODO: full file support for WASI, or an option for it
-// open()
 weak int __syscall_openat(int dirfd, intptr_t path, int flags, ...) {
-  if (!strcmp((const char*)path, "/dev/stdin")) {
+  const char* resolved_path = (const char*)path;
+
+  if (!strcmp(resolved_path, "/dev/stdin")) {
     return STDIN_FILENO;
   }
-  if (!strcmp((const char*)path, "/dev/stdout")) {
+  if (!strcmp(resolved_path, "/dev/stdout")) {
     return STDOUT_FILENO;
   }
-  if (!strcmp((const char*)path, "/dev/stderr")) {
+  if (!strcmp(resolved_path, "/dev/stderr")) {
     return STDERR_FILENO;
   }
-  return -EPERM;
+
+  if (!__paths_resolve_path(&dirfd, &resolved_path)) {
+    return -ENOENT;
+  }
+
+  // Compute rights corresponding with the access modes provided.
+  // Attempt to obtain all rights, except the ones that contradict the
+  // access mode provided to openat().
+  __wasi_rights_t max =
+    ~(__WASI_RIGHTS_FD_DATASYNC | __WASI_RIGHTS_FD_READ |
+      __WASI_RIGHTS_FD_WRITE | __WASI_RIGHTS_FD_ALLOCATE |
+      __WASI_RIGHTS_FD_READDIR | __WASI_RIGHTS_FD_FILESTAT_SET_SIZE);
+  {
+    int accmode = flags & O_ACCMODE;
+    if (accmode == O_RDONLY || accmode == O_RDWR || accmode == O_WRONLY) {
+      if (accmode == O_RDONLY || accmode == O_RDWR) {
+        max |= __WASI_RIGHTS_FD_READ | __WASI_RIGHTS_FD_READDIR;
+      }
+      if (accmode == O_WRONLY || accmode == O_RDWR) {
+        max |= __WASI_RIGHTS_FD_DATASYNC | __WASI_RIGHTS_FD_WRITE |
+               __WASI_RIGHTS_FD_ALLOCATE |
+               __WASI_RIGHTS_FD_FILESTAT_SET_SIZE;
+      }
+    } else if (accmode == O_EXEC || accmode == O_SEARCH) {
+      // Do nothing.
+    } else {
+      return -EINVAL;
+    }
+  }
```

Comment:
Is there some reason to add this extra scope here?

Reply:
Yes, it is there to limit `accmode`'s scope. I don't feel strongly about this, though.